fish-rust/src/builtins/string.rs

```diff
@@ -0,0 +1,493 @@
+                    "Modification of read-only variable \"%ls\" is not allowed\n",
+                    name
+                ));
+            }
+            InvalidEscape(pattern) => {
+                string_error!(
+                    streams,
+                    "%ls: Invalid escape sequence in pattern \"%ls\"\n",
+                    cmd,
+                    pattern
+                );
+            }
+        }
+    }
+}
+
+impl From<crate::wutil::wcstoi::Error> for StringError {
+    fn from(_: crate::wutil::wcstoi::Error) -> Self {
+        StringError::NotANumber
+    }
+}
+
+impl From<std::num::TryFromIntError> for StringError {
+    fn from(_: std::num::TryFromIntError) -> Self {
+        // the result of `try_into` failing on the result of fish_wcsto{i,l}
+        StringError::InvalidValue
+    }
+}
+
+impl From<&'static wstr> for StringError {
+    fn from(s: &'static wstr) -> Self {
+        StringError::InvalidArgs(s)
+    }
+}
+
+impl StringError {
+    fn print_error(
+        &self,
+        args: &[&wstr],
+        parser: &mut parser_t,
+        streams: &mut io_streams_t,
+        optarg: Option<&wstr>,
+        optind: usize,
+        option_name: &wstr,
+    ) {
+        let cmd = args[0];
+        use StringError::*;
+        match self {
+            InvalidArgs(_) | InvalidValue => {
+                let msg = match *self {
+                    InvalidArgs(s) => s.to_owned(),
+                    _ => sprintf!("Invalid %s value", option_name),
+                };
+                let error_msg = wgettext_fmt!("%ls: %ls '%ls'\n", cmd, msg, optarg.unwrap());
+
+                streams.err.append(L!("string "));
+                streams.err.append(error_msg);
+            }
+            NotANumber => {
+                string_error!(streams, BUILTIN_ERR_NOT_NUMBER, cmd, optarg.unwrap());
+            }
+            UnknownOption => {
+                string_unknown_option(parser, streams, cmd, args[optind - 1]);
+            }
+        }
+    }
+
+    fn retval(&self) -> Option<c_int> {
+        STATUS_INVALID_ARGS
+    }
+}
+
+#[derive(Default, PartialEq, Clone, Copy)]
+enum Direction {
+    #[default]
+    Left,
+    Right,
+}
+
+pub(self) fn width_without_escapes(ins: &wstr, start_pos: usize) -> usize {
+    let mut width = 0i32;
+    for c in ins[start_pos..].chars() {
+        let w = fish_wcwidth_visible(c);
+        // We assume that this string is on its own line,
+        // in which case a backslash can't bring us below 0.
+        if w > 0 || width > 0 {
+            width += w;
+        }
+    }
+    // ANSI escape sequences like \e\[31m contain printable characters. Subtract their width
+    // because they are not rendered.
+    let mut pos = start_pos;
+    while let Some(ec_pos) = ins.slice_from(pos).find_char('\x1B') {
+        pos += ec_pos;
+        if let Some(len) = escape_code_length(ins.slice_from(pos)) {
+            let sub = ins.slice_from(pos).slice_to(len);
```

Comment:
> I vote for using bracket indexing for the sake of consistency

I agree, but then we should probably just remove `slice_{to,from}`

```suggestion
            let sub = &ins[pos..len];
```

Reply:
> I agree, but then we should probably just remove slice_{to,from} from WExt

yeah. I'm not sure what others think

Reply:
The logical distinction of `slice_to` and `slice_from` is that they operate on char counts, instead of byte counts.

In a hypothetical future where we try to switch to UTF-8, that distinction will become important. I worry that the bracket indexing will introduce bugs at that point. But I'm open to other ideas for how to handle that.

Reply:
Interesting; a clippy lint banning str slicing seems easy and robust enough